core/client/fs/src/main/java/alluxio/client/block/stream/SharedGrpcDataReader.java

```diff
@@ -0,0 +1,190 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.client.block.stream;
+
+import alluxio.client.file.FileSystemContext;
+import alluxio.grpc.ReadRequest;
+import alluxio.network.protocol.databuffer.DataBuffer;
+import alluxio.network.protocol.databuffer.NioDataBuffer;
+import alluxio.resource.LockResource;
+import alluxio.wire.WorkerNetAddress;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * A shared gRPC data reader that cache blocks data for multi-thread accessing.
+ *
+ * It follows GrpcDataReader protocol and takes strong assumption:
+ * Parallel read to the same file happens on the same time, so that read request is
+ * serialized by kernel
+ */
+@NotThreadSafe
+public class SharedGrpcDataReader implements DataReader {
+  private static final Logger LOG = LoggerFactory.getLogger(SharedGrpcDataReader.class);
+  private static final int BLOCK_LOCK_NUM = 32;
```

Comment:
Would it be useful for this to be configurable?

Reply:
I don't think this value needs to be configured